include/zephyr/sys_clock.h

```diff
@@ -190,7 +190,78 @@ int64_t sys_clock_tick_get(void);
 #define sys_clock_tick_get_32() (0)
 #endif
 
-uint64_t sys_clock_timeout_end_calc(k_timeout_t timeout);
+/**
+ * @brief Kernel timepoint type
+ *
+ * Absolute timepoints are stored in this opaque type.
+ * It is best not to inspect its content directly.
+ *
+ * @see sys_time_calc()
+ * @see sys_time_remaining()
+ * @see sys_time_expired()
+ */
+typedef struct { uint64_t tick; } k_timepoint_t;
+
+/**
+ * @brief Calculate a timepoint value
+ *
+ * Returns a timepoint corresponding to the expiration (relative to an
+ * unlocked "now"!) of a timeout object.  When used correctly, this should
+ * be called once, synchronously with the user passing a new timeout value.
+ * It should not be used iteratively to adjust a timeout (see
+ * `sys_time_remaining()` for that purpose).
+ *
+ * @param timeout Timeout value relative to current time (may also be
+ *                `K_FOREVER` or `K_NO_WAIT`).
+ * @retval Timepoint value corresponding to given timeout
+ *
+ * @see sys_time_remaining()
+ * @see sys_time_expired()
+ */
+k_timepoint_t sys_time_calc(k_timeout_t timeout);
```

Comment:
Naming nitpicks: "sys_time" seems a little ambiguous.  This is expressing a property of a timeout object, so I'd argue it should say "timeout".  The original use of "calc" was (I guess?  It was a while ago and I don't remember) intended to convey that this was doing math based on current state, but as pointed out here that wasn't really clear either.

How about "sys_timeout_expires()" or "_expiration()" instead?  It's the expire time of a timeout, expressed in the newly minted data type intended for that purpose.

Reply:
Well, I was envisioning this as a concept to represent timepoints. As such, this is not a timeout property per se but rather that a timeout object is used to create a timepoint object. Timepoints could possibly be created based on something else than a timeout eventually. And the «timeout» namespace is already quite loaded already. Timepoints are absolute while timeouts are relative. I wanted to avoid confusing them by not having the same namespace. Does this make any sense?
This being said I'm all for better naming if we can find one. I initially had sys_timepoint_* but I thought sys_time_* looked better.

Reply:
I'd say keep with the `sys_timepoint_` namespace as that gives you a green field to play in, and makes it clear that they're working in this new timespace. Maybe `sys_timespace_from_timeout` in this case? Naming is almost as hard as off-by-one errors.